# Исследование базы данных

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Цель исследования - анализ базы данных в рамках описанных задач

## Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

In [29]:
#импортируем библиотеки
import pandas as pd 
from sqlalchemy import create_engine 

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
 db_config['pwd'], 
 db_config['host'], 
 db_config['port'], 
 db_config['db']) 

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

Выведем первые 10 строк каждой таблицы

In [30]:
def extract(query):
    display(pd.io.sql.read_sql(query, con = engine)) 

In [31]:
extract('SELECT * FROM books LIMIT 10')


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

In [32]:
extract("SELECT * FROM authors LIMIT 10")

author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

In [33]:
extract("SELECT * FROM publishers LIMIT 10")

publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

In [34]:
extract("SELECT * FROM ratings LIMIT 10")

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

In [35]:
extract("SELECT * FROM reviews LIMIT 10")

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

1. Посчитайте, сколько книг вышло после 1 января 2000 года

In [36]:
extract("SELECT COUNT(book_id) FROM books")
extract("SELECT COUNT(book_id) FROM books WHERE publication_date > '2000-01-01'")

count
0   1000

count
0    819

Всего в базе 1000 книг, из них 819 вышло после 1 января 2000 года

2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [40]:
extract('''
WITH reviews AS (
SELECT b.book_id, title, COUNT(review_id) reviews
FROM books b 
FULL JOIN reviews r ON b.book_id = r.book_id
GROUP BY b.book_id 
), 
rating AS (
SELECT b.book_id, title, AVG(rating) avg_rating 
FROM books b 
FULL JOIN ratings rat ON b.book_id = rat.book_id 
GROUP BY b.book_id 
)

SELECT reviews.title, reviews, avg_rating FROM
reviews
FULL JOIN rating ON reviews.book_id = rating.book_id
ORDER by 3 DESC
''')

title  reviews  avg_rating
0        Arrows of the Queen (Heralds of Valdemar  #1)        2        5.00
1    The Walking Dead  Book One (The Walking Dead #...        2        5.00
2                                      Light in August        2        5.00
3    Wherever You Go  There You Are: Mindfulness Me...        2        5.00
4    Captivating: Unveiling the Mystery of a Woman'...        2        5.00
..                                                 ...      ...         ...
995  The World Is Flat: A Brief History of the Twen...        3        2.25
996                  His Excellency: George Washington        2        2.00
997                                      Drowning Ruth        3        2.00
998                                              Junky        2        2.00
999                               Harvesting the Heart        2        1.50

[1000 rows x 3 columns]

Вывод: книга с высоким рейтингом не всегда провоцирует писать много обзоров.

3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [26]:
extract('''
SELECT publisher, COUNT(book_id) 
FROM books b 
FULL JOIN publishers p ON b.publisher_id = p.publisher_id
WHERE num_pages > 50
GROUP BY p.publisher_id
ORDER BY 2 DESC
LIMIT 1
''')

publisher  count
0  Penguin Books     42

Издательство Penguin Books выпустило 42 книги из имеющихся в базе толще 50 страниц

4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [27]:
extract('''
SELECT author, AVG(rating) 
FROM books b 
JOIN authors a ON b.author_id = a.author_id
JOIN ratings r ON b.book_id = r.book_id
GROUP BY author
HAVING COUNT(r.rating_id) > 50
ORDER BY 2 desc
LIMIT 1
''')

author       avg
0  J.K. Rowling/Mary GrandPré  4.288462

Побеждает J.K. Rowling - автор книг о Гарри Поттере со средней оценкой 4.29

5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [28]:
extract('''
SELECT AVG(count)
FROM(
    SELECT username, COUNT(review_id)
    FROM reviews
    WHERE username in (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(rating_id) > 50
    )
    GROUP BY username) t
''')

avg
0  24.333333

Около 24 обзоров написал в среднем каждый, кто поставил более 50 оценок

# Общий вывод

Предоставленная база данных сервиса работоспособна и подходит для получения ответов на вопросы в рамках поставленных задач.